nav2_smac_planner/src/a_star.cpp

```diff
@@ -229,15 +290,23 @@ bool AStarAlgorithm<NodeT>::areInputsValid()
   }
 
   // Check if points were filled in
-  if (!_start || !_goal) {
+  if (!_start || _goalsSet.empty()) {
     throw std::runtime_error("Failed to compute path, no valid start or goal given.");
   }
 
   // Check if ending point is valid
-  if (getToleranceHeuristic() < 0.001 &&
-    !_goal->isNodeValid(_traverse_unknown, _collision_checker))
-  {
-    throw nav2_core::GoalOccupied("Goal was in lethal cost");
+  if (getToleranceHeuristic() < 0.001) {
```

Comment:
What about the goal coordinates? 

After this step there will be a mismatch better the goalSet and the goal_coordinates.

Reply:
Yes, I did that on purpose because I use the goal_coordinates for comparision when i call the setgoals function and I dont want to lose the order  because i need it here getInitialGoalCoordinate.

Reply:
Yeah so the current flow is 

1. setGoal get called, initializes the goalSet and goals_coords
2. CreatePath -> areInputsValid which adjusts goalSet, Note: this could invalidate the first goals_coord value i.e  (goals_coords[0]). After this point goalSet.size() != goals_coords.size()
3. Then initial coord get passing to try analytic expansion, the heading of this coord can be invalid. 
In the analytic expansion, getHeuristicCost is calucalted _which_ is a function of the heading. 

As a result, you might up attempting analytic expansions in the wrong iterations. You need to calculate the heuristic cost of the valid set of all goal and take the min.

Reply:
You're correct. The change I implemented, which includes pruning, has led to additional problems. In addition to what you said, during the initial stage, I calculate the distance heuristic in advance based on the goal heading with the lowest cost, but there's a chance this goal might be invalid and the heuristic can led the wrong goal. The optimal solution would be to adjust the distance heuristic lookup to depend on the heading. I plan to apply this only for standard and bidirectional headings, considering the impact of the distance heuristic is relatively minor compared to the obstacle heuristic, what do you think?

Reply:
I think you should also do it for the all direction headings. 

The easiest thing would be to create a set of lookup tables rather than just taking the min. Then based on the goals used you construct the final min lookup table used for the search.

Reply:
@stevedanomodolor were these addresses?